bittide-instances/tests/Wishbone/CaptureUgn.hs

```diff
@@ -0,0 +1,146 @@
+-- SPDX-FileCopyrightText: 2024 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# OPTIONS_GHC -fplugin=Protocols.Plugin #-}
+
+module Wishbone.CaptureUgn where
+
+import Clash.Explicit.Prelude
+import Clash.Prelude (withClockResetEnable)
+
+import Clash.Cores.UART (ValidBaud, uart)
+import Clash.Signal.Internal
+import Data.Char
+import Data.Maybe
+import Language.Haskell.TH
+import Numeric
+import Project.FilePath
+import Protocols
+import System.FilePath
+import Test.Tasty
+import Test.Tasty.HUnit
+import Test.Tasty.TH
+import VexRiscv
+
+import Bittide.CaptureUgn
+import Bittide.DoubleBufferedRam
+import Bittide.Instances.Domains
+import Bittide.ProcessingElement
+import Bittide.ProcessingElement.Util
+import Bittide.SharedTypes
+import Bittide.Wishbone
+import Clash.Cores.UART.Extra (MaxBaudRate)
+
+import qualified Prelude as P
+
+{- | Test whether we can read the local and remote sequence counters from the captureUgn
+peripheral.
+-}
+case_capture_ugn_self_test :: Assertion
+case_capture_ugn_self_test =
+  assertBool
+    msg
+    ( actualLocalCounter
+        == expectedLocalCounter
+        && actualRemoteCounter
+        == expectedRemoteCounter
+    )
+ where
+  msg =
+    "Received local counter 0x"
+      <> showHex actualLocalCounter ""
+      <> " and remote counter 0x"
+      <> showHex actualRemoteCounter ""
+      <> " are not equal to expected counters 0x"
+      <> showHex expectedLocalCounter ""
+      <> " and 0x"
+      <> showHex expectedRemoteCounter ""
+  (expectedLocalCounter, unpack -> expectedRemoteCounter) = getSequenceCounters $ bundle (localCounter, eb)
+  (actualLocalCounter, actualRemoteCounter) = parseResult simResult
+  baud = SNat @(MaxBaudRate Basic50)
+  clk = clockGen
+  rst = resetGen
+  ena = enableGen
+  simResult = fmap (chr . fromIntegral) $ catMaybes $ sampleN 100_000 uartStream
+  (uartStream, _, _) = withClockResetEnable (clockGen @Basic50) rst ena $ uart baud uartTx (pure Nothing)
+  uartTx = dut baud clk rst eb localCounter (pure 0)
+
+  {- The local counter starts counting up from 0x1122334411223344. The elastic buffer
+  outputs Nothing for 1000 cycles, after which it will  start outputting a decreasing
+  counter starting at 0xaabbccddeeff1234.
+  -}
+  localCounter = register clk rst ena 0xaabbccddeeff1234 (localCounter + 1)
+
+  eb = regEn clk rst ena Nothing remoteStarted (Just . pack <$> remoteCounter)
+   where
+    remoteCounter :: Signal Basic50 (Unsigned 64)
+    remoteCounter = register clk rst ena 0x1122334411223344 (remoteCounter - 1)
+    remoteStarted = counter .==. pure maxBound
+    counter :: Signal Basic50 (Index 1000)
+    counter = register clk rst ena 0 (satSucc SatBound <$> counter)
```

Comment:
You can move the `Index 1000` annotation to the initial value and remove the `counter` type annotation.